app/services/tag_set_service.rb

```diff
@@ -0,0 +1,21 @@
+# frozen_string_literal: true
+
+# TagSetService
+# A service for fetching whole tag sets into memory
+class TagSetService
+  attr_reader :loaded_tag_sets
+
+  def initialize
+    @loaded_tag_sets = {}
+  end
+
+  def load_tag_set(tag_set_name)
+    return if loaded_tag_sets.key?(tag_set_name)
```

Comment:
I would expect this would be an error in the workflow to receive anything but the correct key.  Maybe it should cause an exception rather than just overwriting the loaded tag set.